### DISCUSSION TITLE
Twitch - Records only one seconds with API

### DISCUSSION THREAD

**Discussion by tomy137** [streamlink/streamlink#3516] [Q&A] [answered]
Hello there OWK, 

I am struggling with this code : 
```
import streamlink
from streamlink import Streamlink

session = Streamlink()
streams = session.streams("https://www.twitch.tv/lestream")

f = open('new-video.mp4', "wb")
f.write(streams["best"].open().read(-1))
f.close()
```

The recording stop few seconds after begginning and just one or two seconds of the video is saved. 

I've tested directly with the command line client and it's working fine. 
What I am doing wrong ? 

Thnaks in advance for your help !

**Comment by bastimeyer** [comment, marked as answer]
This is not how the stream's read method words. See [`SegmentedStreamReader.read`](https://github.com/streamlink/streamlink/blob/2.0.0/src/streamlink/stream/segmented.py#L224-L229) / [`RingBuffer.read`](https://github.com/streamlink/streamlink/blob/2.0.0/src/streamlink/buffers.py#L104-L112).

If you only call `read()` once, then it'll block the current thread (unless `block=False`) until the `SegmentedStreamWriter` has written data into the stream's buffer. Once data is available, the blocked read method will return that data. This means that you'll have to loop the read call until the stream ends, which is indicated by an empty bytes object. Or until your program's execution is supposed to stop. How you loop it is up to you.

**Reply by tomy137** [reply]
Ok, thanks ! For example, this code is working fine : 

```
import streamlink
from streamlink import Streamlink

session = Streamlink()
streams = session.streams("https://www.twitch.tv/kaellyn")
s = streams["best"].open()

f = open('new-video.mp4', "wb")

for i in range(0,20) :  # Just for testing
  f.write(s.read(-1))
f.close()
```

**Reply by bastimeyer** [reply]
See [`builtins.iter`](https://docs.python.org/3/library/functions.html#iter) and [`functools.partial`](https://docs.python.org/3/library/functools.html#functools.partial):
```py
iter(partial(stream.read, chunk_size), b"")
```

Examples:
- https://github.com/streamlink/streamlink/blob/a03f7bfc6932cacfc7a89e62ff509c0be7294f51/src/streamlink_cli/main.py#L347-L385
- https://github.com/streamlink/streamlink/blob/a03f7bfc6932cacfc7a89e62ff509c0be7294f51/tests/mixins/stream_hls.py#L151-L153
